k9mail/src/main/java/com/fsck/k9/helper/ReplyToParser.java

```diff
@@ -1,23 +1,78 @@
 package com.fsck.k9.helper;
 
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import com.fsck.k9.Account;
 import com.fsck.k9.mail.Address;
 import com.fsck.k9.mail.Message;
+import com.fsck.k9.mail.Message.RecipientType;
 import com.fsck.k9.mail.internet.ListHeaders;
 
 
 public class ReplyToParser {
-    public static Address[] getRecipientsToReplyTo(Message message) {
-        Address[] replyToAddresses = message.getReplyTo();
-        if (replyToAddresses.length > 0) {
-            return replyToAddresses;
+
+    public static Address[] getRecipientsToReplyTo(Message message, Account account) {
+        Address[] result = message.getReplyTo();
+
+        boolean hasNoReplyToAddresses = (result.length == 0);
+        if (hasNoReplyToAddresses) {
+            result = ListHeaders.getListPostAddresses(message);
+        }
+
+        boolean hasNoListPostAddresses = result.length == 0;
+        if (hasNoListPostAddresses) {
+            result = message.getFrom();
         }
 
-        Address[] listPostAddresses = ListHeaders.getListPostAddresses(message);
-        if (listPostAddresses.length > 0) {
-            return listPostAddresses;
+        boolean replyToAddressIsUserIdentity = account.isAnIdentity(result);
+        if (replyToAddressIsUserIdentity) {
+            result = message.getRecipients(RecipientType.TO);
         }
 
-        return message.getFrom();
+        return result;
     }
+
+    public static ReplyToAddresses getRecipientsToReplyAllTo(
+            Message message, Address[] replyToAddresses, Account account) {
+        ArrayList<Address> alreadyAddedAddresses = new ArrayList<>(Arrays.asList(replyToAddresses));
+        ArrayList<Address> toAddresses = new ArrayList<>();
+        ArrayList<Address> ccAddresses = new ArrayList<>();
+
+        for (Address address : message.getFrom()) {
+            if (!account.isAnIdentity(address) && !alreadyAddedAddresses.contains(address)) {
```

Comment:
Switching the order of the conditions avoids calling the more expensive `isAnIdentity()` for addresses we have already added.